pkgs/stdenv/generic/make-derivation.nix

```diff
@@ -542,11 +548,37 @@ else let
         "The ‘env’ attribute set can only contain derivation, string, boolean or integer attributes. The ‘${n}’ attribute is of type ${builtins.typeOf v}."; v)
       env;
 
+  drvAttrs = derivationArg // lib.optionalAttrs envIsExportable checkedEnv;
+  strict = builtins.derivationStrict drvAttrs;
+  outputName = lib.head outputs;
+
+  name = attrs.name or attrs.pname;
+
+  # begin __cleanAttrs helpers
+
+  whatNow =
+    (
+      if meta.maintainers != []
+      then "You may work with the maintainer(s) of ${name}: ${lib.concatMapStringsSep ", " (m: m.github or m.name or m.email) meta.maintainers} (and/or the community), to add explicit support for your use case."
+      else "You may work with the community to add explicit support for your use case to ${name}."
+    ) +
+    " See https://nixos.org/manual/nixpkgs/unstable/#warning-package-attr-impl-detail"
+    ;
+
+  warnCleanAttrs = lib.mapAttrs (k: lib.warn "The attribute ${lib.strings.escapeNixIdentifier k} of package ${name} ${if k == "passthru" then "is" else "seems to be"} an implementation detail and may be removed from the package attribute set in the future. ${whatNow}");
+
+  warnCleanDrvAttrs = lib.warn "The attribute drvAttrs of package ${name} is an implementation detail and may be removed from the package attribute set in the future. ${whatNow}";
+
+  maybeWarnCleanAttrs = if __cleanAttrs == "warn" then warnCleanAttrs else x: x;
+
+  # end __cleanAttrs helpers
+
 in
 
 lib.extendDerivation
   validity.handled
-  ({
+  (
+    lib.optionalAttrs (__cleanAttrs != true) (maybeWarnCleanAttrs {
```

Comment:
This removes `inputDerivation` when `__cleanAttrs == true`, I'm not sure if that makes sense, since `.inputDerivation` is meant to be a public stable interface. While it might not be needed anymore since [`mkShell` is now buildable](https://github.com/NixOS/nixpkgs/pull/153194), I'm not sure if this should be decided here.